docs/technical-design/contract-rate-refactor-relationships.md

```diff
@@ -0,0 +1,163 @@
+
+## Contract and Rate Revisions tables
+
+Just as we have maintained a list of package revisions up until now, this migration will entail the creation of a rates revision table. Every time a rate's info is updated, we will create a new entry in the rate revisions table with the new version of the data. This way a history of all past versions of the rate is maintained as a list of revisions. 
+
+The trick is in tracking the history of the relationships between contract revisions and rate revisions. Since we only want to add a new entry to a revision table when the underlying form data changes, that means we need to track the relationships outside of those revision tables, and we need to be able to track a changing set of relationships between otherwise unchanged revisions over time. 
+
+The plan is to have a join table between ContractRevisionTable and RateRevisionTable that in addition to the expected `contract_revision_id` and `rate_revision_id` fields will have fields `valid_after`, `valid_before` timestamps and a `isRemoval` boolean. Whenever the relationships between different contracts and rates are changed, we will add new entries to this join table to record the current state. With those dates recorded, it will be possible to scan the join table for a given revision and construct a history of its relationships to other revisions. By including an entry in the join table indicating when a relationship is removed, it makes it straightforward to include that action as an entry in the history as well.
+
+
+### Draft Revisions
+
+Before submission, state users will fill out a draft revision. Draft revisions are easily identifiable as having no submitInfo associated with them. Draft revisions do not exist in the join table, they track which Rates or Contracts they are associated with using the draftRates or draftContracts relationship. We only add to the revision join table when a revision is submitted, recording the relationships for posterity. The separate table means that if your draft contract is created being tied to a rate that is re-submitted, then our draft doesn’t have an outdated link to the old revision. We can always pull the most recent submitted revision when talking about the draft.
+
+
+### The Database Functions
+
+The contract and rate lifecycles are implemented in our postgres package. There are four functions each for Contracts and Rates for manipulating them:
+
+```
+insertDraftContract
+updateDraftContract
+submitContract
+unlockContract
+```
+
+Plus a couple functions for inspecting them:
+
+```
+findContractWithHistory
+findDraftContract
+```
+
+The lifecycle functions are used as follows
```

Comment:
I like this section! A couple suggestions:
 1.  Discuss t he rate functions discussed here too, not just the contract functions. Are the rates functions just the reverse of the contract functions? Are there any differences? This is fuzzy here because the rate functions aren't discussed. 
 2.  Would like to see `findContractWithHistory` and `findDraftContract` each have a one sentence description.  It seems like they are supposed to be captured in the "one gotcha here" paragraph but its not very explicit how that relates to the functions themselves.